File: src/com/easy/FindTheDifference_389.java
/**
 * 
 */
package com.easy;

/**
 * @author Chris Lin
 * @version 1.0
 */
public class FindTheDifference_389 {
    // 2019年7月17日
    // Use Array count
    // Time Complexity: O(n)
    // Space Complexity:O(1)
    // Runtime: 2 ms, faster than 45.17%
    // Memory Usage: 37.7 MB, less than 7.13%
    public char findTheDifference(String s, String t) {
        int[] mCount = new int[26];

        for (int i = 0; i < s.length(); i++) {
            mCount[s.charAt(i) - 'a']++;
        }
        for (int i = 0; i < t.length(); i++) {
            if (--mCount[t.charAt(i) - 'a'] < 0) {
                return t.charAt(i);
            }
        }
        return 0;
    }
}

File: Golang_src/easy/findTheDifference_389.go
package easy

// Use hashmap:
// Time Complexity: O(n)
// Space Complexity:O(|Σ|)
// for |Σ| = 26
// Runtime: 0 ms, faster than 100.00%
// Memory Usage: 2.1 MB, less than 95.41%
func findTheDifference(s string, t string) byte {
	cnt := make([]int, 26)

	for _, v := range s {
		cnt[v-'a']++
	}

	for _, v := range t {
		cnt[v-'a']--
		if cnt[v-'a'] < 0 {
			return byte(v)
		}
	}

	return ' '
}

// Use XOR:
// Time Complexity: O(n)
// Space Complexity:O(1)
// Runtime: 3 ms, faster than 42.20%
// Memory Usage: 2.1 MB, less than 95.41%
func findTheDifference2(s string, t string) byte {
	var diff byte

	for i := range s {
		diff ^= s[i] ^ t[i]
	}

	diff ^= t[len(t)-1]

	return diff
}
